### PR TITLE
Expose TilesetDescriptorFactory

### LINKED ISSUE #4282: Expose TilesetDescriptorFactory API
To load OfflineRegions customer should provide `TilesetDescriptors`. 
For `navigation` tiles it can be done with `com.mapbox.navigator.TilesetDescriptorFactory`
SDK should expose API to access the factory ^^.

### PR DESCRIPTION
**Pull request by korshaknn**
### Description

closes #4282 

### Changelog

```
<changelog>Expose TilesetDescriptorFactory</changelog>
```

### CODE REVIEW

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/TilesetDescriptorFactory.kt [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+private typealias NativeTilesetDescriptorFactory = com.mapbox.navigator.TilesetDescriptorFactory
+
+/**
+ * A factory to build navigation [TilesetDescriptor]
+ */
+class TilesetDescriptorFactory internal constructor(
+    navigator: MapboxNativeNavigator
+) {
+    private val cache: CacheHandle = navigator.cache!!
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/TilesetDescriptorFactory.kt:15]: Is this always non-null?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/TilesetDescriptorFactory.kt [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import com.mapbox.common.TilesetDescriptor
+import com.mapbox.navigation.navigator.internal.MapboxNativeNavigator
+import com.mapbox.navigator.CacheHandle
+
+private typealias NativeTilesetDescriptorFactory = com.mapbox.navigator.TilesetDescriptorFactory
+
+/**
+ * A factory to build navigation [TilesetDescriptor]
+ */
+class TilesetDescriptorFactory internal constructor(
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/TilesetDescriptorFactory.kt:12]: Is there an option to expose just:
```kotlin
fun build(): TilesetDescriptor
```

that will take the current options from the `MapboxNavigation`/`Navigator`?

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/PredictiveCacheController.kt [hunk 3]
```diff
@@ -174,4 +162,13 @@ class PredictiveCacheController @JvmOverloads constructor(
...
     }
+
+    private fun MapboxMap.tileStore(): TileStore {
+        val tileStorePath = this.getResourceOptions().tileStorePath
+        return if (tileStorePath == null) {
+            TileStore.getInstance()
+        } else {
+            TileStore.getInstance(tileStorePath)
+        }
+    }
```
Comment (LukasPaczos) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/PredictiveCacheController.kt:173]: How can we make this function as robust to use as possible?

The map and navigation tilestore instances should always have the same path. Should we add logs if that's not the case? Throw exceptions?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/TilesetDescriptorFactory.kt (hunk 1 above)
Comment (korshaknn) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/TilesetDescriptorFactory.kt:15]: yes, must be non-null

Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/TilesetDescriptorFactory.kt:15]: Why is the type nullable in this case?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/TilesetDescriptorFactory.kt (hunk 2 above)
Comment (korshaknn) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/TilesetDescriptorFactory.kt:12]: no, we need 3 functions:
- build descriptor with a passed version+dataset
- build with server version
- build with local version used by navigator

Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/TilesetDescriptorFactory.kt:12]: Why can't we fulfill that local/server dataset name based on the current routing tiles options?

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/PredictiveCacheController.kt (hunk 3 above)
Comment (korshaknn) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/PredictiveCacheController.kt]: removed this changes, rebase issue

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/TilesetDescriptorFactory.kt (hunk 2 above)
Comment (korshaknn) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/TilesetDescriptorFactory.kt:12]: what if customer need another dataset for offline regions?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/TilesetDescriptorFactory.kt (hunk 1 above)
Comment (korshaknn) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/TilesetDescriptorFactory.kt:15]: all of the objects are nullable in `MapboxNativeNavigatorImpl`:
```kotlin
private var navigator: Navigator? = null
    private var nativeRouter: Router? = null
    private var historyRecorderHandle: HistoryRecorderHandle? = null
    private var route: DirectionsRoute? = null
    private var routeBufferGeoJson: Geometry? = null
    override var graphAccessor: GraphAccessor? = null
    override var openLRDecoder: OpenLRDecoder? = null
    override var roadObjectsStore: RoadObjectsStore? = null
    override var cache: CacheHandle? = null
    private var logger: Logger? = null
```
and are inited in `override fun create()` and it must be called before any call to `MapboxNativeNavigatorImpl`

we also have a lot of calls with ` navigator!!`
so I think it's ok to have `navigator.cache!!`

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/TilesetDescriptorFactory.kt (hunk 2 above)
Comment (korshaknn) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/TilesetDescriptorFactory.kt:12]: we can make a `dataset` param nullable and pass null by default. 
if null -> use from options
non null -> use it

**Comment by korshaknn** [comment]
@LukasPaczos 
- changed `cache` to `lateinit non-null`
- use `routingTilesOptions` if no params provided
- added tests

**Review by LukasPaczos — APPROVED** [review]
LGTM :ship:

**Pull request merged**